File: frontend/src/types/response/filesystem.ts
//
export interface CreateFolderResponse extends BaseResponse<{
    status: boolean;
    message: string;
}> {
}

export interface RenameFolderResponse extends BaseResponse<{
    status: boolean;
    message: string;
}> {
}

export interface DeleteFolderResponse extends BaseResponse<{
    status: boolean;
    message: string;
}> {
}

// // File 文件结构体
// type File struct {
//     gorm.Model        // ID, CreatedAt, UpdatedAt, DeletedAt
//     Type       string `gorm:"not null" json:"type"`  // 类型
//     Name       string `gorm:"not null" json:"name"`  // 文件名称
//     Key        string `gorm:"not null" json:"key"`   // 文件唯一标识 UUID_V4{9,23}
//     Level      uint   `gorm:"not null" json:"level"` // 文件层级
//     FolderID   uint
//     Folder     Folder `gorm:"foreignkey:FolderID"` // 所属文件夹对象
// }
//
// // Folder 文件夹结构体
// type Folder struct {
//     gorm.Model          // ID, CreatedAt, UpdatedAt, DeletedAt
//     Type       string   `gorm:"not null" json:"type"`     // 类型
//     Name       string   `gorm:"non null" json:"name"`     // 文件夹名
//     Key        string   `gorm:"not null" json:"key"`      // 文件夹唯一标识 UUID_V4{9,23}
//     Level      uint     `gorm:"not null" json:"level"`    // 文件夹层级
//     ParentID   uint     `gorm:"not null" json:"parentID"` // // 父文件夹ID，根目录的父文件夹ID设为0
//     Parent     *Folder  `gorm:"foreignkey:ParentID"`      // 父文件夹对象
//     Subfolders []Folder `gorm:"foreignkey:ParentID"`      // 子文件夹列表
//     Files      []File   // 文件列表
// }
//
// // Storage 文件系统根节点结构体
// type Storage struct {
//     Root Folder `json:"root"`
// }

export interface File {
    ID: number;
    type: string;
    name: string;
    key: string;
    level: number;
    folderID: number;
}

export interface Folder {
    ID: number;
    type: string;
    name: string;
    key: string;
    level: number;
    parentID: number;
    parent: Folder | null;
    subfolders: Array<Folder> | null;
    files: Array<File> | null;
}

export interface GetStorageResponse extends BaseResponse<{
    status: boolean;
    message: string;
    storage: Array<Folder>;
}> {
}

File: backend/controller/filesystem.go
package controller

import (
	"HostManager/backend/model"
	"HostManager/backend/service"
)

// FileSystem 文件系统API
type FileSystem struct {
	Base
	ser *service.FileSystemSrv
}

// NewFileSystem 创建FileSystem
func NewFileSystem() *FileSystem {
	return &FileSystem{
		ser: service.GetFileSystemSrv(),
	}
}

func (f *FileSystem) GetStorage() Res {
	type GetStorageResponse struct {
		Status  bool           `json:"status"`
		Message string         `json:"message"`
		Storage []model.Folder `json:"storage"`
	}
	//
	storage, err := f.ser.GetStorage()
	if err != nil && err.Error() != "record not found" {
		return f.success(GetStorageResponse{Status: false, Message: err.Error()})
	}
	return f.success(GetStorageResponse{Status: true, Message: "获取文件系统成功", Storage: storage})
}

// CreateFolder 创建目录
func (f *FileSystem) CreateFolder(name string, parentID uint64) Res {
	type CreateFolderResponse struct {
		Status  bool   `json:"status"`
		Message string `json:"message"`
	}
	//
	if name == "" || parentID < 0 {
		return f.success(CreateFolderResponse{Status: false, Message: "目录参数错误"})
	}
	//
	err := f.ser.CreateFolder(name, parentID)

	if err != nil {
		return f.success(CreateFolderResponse{Status: false, Message: err.Error()})
	}
	return f.success(CreateFolderResponse{Status: true, Message: "创建目录成功"})
}

// RenameFolder 重命名目录
func (f *FileSystem) RenameFolder(name string, id uint64) Res {
	type RenameFolderResponse struct {
		Status  bool   `json:"status"`
		Message string `json:"message"`
	}
	//
	if id < 0 || name == "" {
		return f.success(RenameFolderResponse{Status: false, Message: "目录参数错误"})
	}
	//
	err := f.ser.RenameFolder(name, id)
	if err != nil {
		return f.success(RenameFolderResponse{Status: false, Message: err.Error()})
	}
	return f.success(RenameFolderResponse{Status: true, Message: "重命名目录成功"})
}

// DeleteFolder 删除文件夹
func (f *FileSystem) DeleteFolder(id uint64) Res {
	type DeleteFolderResponse struct {
		Status  bool   `json:"status"`
		Message string `json:"message"`
	}
	//
	if id < 0 {
		return f.success(DeleteFolderResponse{Status: false, Message: "目录参数错误"})
	}
	//
	err := f.ser.DeleteFolder(id)
	if err != nil {
		return f.success(DeleteFolderResponse{Status: false, Message: err.Error()})
	}
	return f.success(DeleteFolderResponse{Status: true, Message: "删除目录成功"})
}
